Move the StorageVersions types to kube-apiserver

roycaihw opened the pull request:
/kind api-change

Move the internal types to kube-apiserver, and the external types to k8s.io/api. Ref https://github.com/kubernetes/kubernetes/pull/92064#discussion_r447795942

The first three commits are based on https://github.com/kubernetes/kubernetes/pull/93164

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

/sig api-machinery
/priority important-soon
/assign @caesarxuchao

roycaihw committed:
move apiserverinternal types to kube-apiserver

move versioned types to k8s.io/api;
cleanup generated files

caesarxuchao commented on staging/src/k8s.io/api/apiserverinternal/v1alpha1/doc.go:
```diff
@@ -16,12 +16,8 @@ limitations under the License.
 // +k8s:openapi-gen=true
-// +k8s:defaulter-gen=TypeMeta
-
-// +groupName=internal.apiserver.k8s.io
```
I expect this line is kept, like https://github.com/kubernetes/kubernetes/blob/adefb2ebc8866bac3b17a6e7bbdf33a31128ec3d/staging/src/k8s.io/api/flowcontrol/v1alpha1/doc.go#L21

caesarxuchao commented on staging/src/k8s.io/client-go/informers/generic.go:
```diff
@@ -100,6 +101,10 @@ func (f *sharedInformerFactory) ForResource(resource schema.GroupVersionResource
 	case v1beta1.SchemeGroupVersion.WithResource("validatingwebhookconfigurations"):
 		return &genericInformer{resource: resource.GroupResource(), informer: f.Admissionregistration().V1beta1().ValidatingWebhookConfigurations().Informer()}, nil
 
+		// Group=apiserverinternal, Version=v1alpha1
```
`Group=apiserverinternal` is wrong.
I think if you keep the groupName tag, the comment here would be correct.

caesarxuchao commented on staging/src/k8s.io/client-go/kubernetes/typed/apiserverinternal/v1alpha1/fake/fake_storageversion.go:
```diff
@@ -0,0 +1,133 @@
+	Fake *FakeApiserverinternalV1alpha1
+}
+
+var storageversionsResource = schema.GroupVersionResource{Group: "apiserverinternal", Version: "v1alpha1", Resource: "storageversions"}
```
The group here is wrong.

caesarxuchao commented on staging/src/k8s.io/client-go/kubernetes/typed/apiserverinternal/v1alpha1/fake/fake_storageversion.go:
```diff
@@ -0,0 +1,133 @@
+
+var storageversionsResource = schema.GroupVersionResource{Group: "apiserverinternal", Version: "v1alpha1", Resource: "storageversions"}
+
+var storageversionsKind = schema.GroupVersionKind{Group: "apiserverinternal", Version: "v1alpha1", Kind: "StorageVersion"}
```
The Group here is wrong.

caesarxuchao reviewed (commented):
Otherwise lgtm.

roycaihw commented:
/retest

roycaihw commented:
/retest

roycaihw committed:
update doc.go and register.go

roycaihw committed:
add internal.apiserver.k8s.io/v1alpha1 to known gvs

roycaihw committed:
fix import paths

roycaihw committed:
generated

roycaihw commented:
/retest

caesarxuchao commented:
A nit, otherwise lgtm.

caesarxuchao commented:
/lgtm

roycaihw commented:
/assign @liggitt @lavalamp

roycaihw commented:
Gentle ping

liggitt commented:
add to:
```
diff --git a/staging/src/k8s.io/api/roundtrip_test.go b/staging/src/k8s.io/api/roundtrip_test.go
index beafa9d1a25..0bc318a5da4 100644
--- a/staging/src/k8s.io/api/roundtrip_test.go
+++ b/staging/src/k8s.io/api/roundtrip_test.go
@@ -24,6 +24,7 @@ import (
 	admissionv1beta1 "k8s.io/api/admission/v1beta1"
 	admissionregv1 "k8s.io/api/admissionregistration/v1"
 	admissionregv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	apiserverinternalv1alpha1 "k8s.io/api/apiserverinternal/v1alpha1"
 	appsv1 "k8s.io/api/apps/v1"
 	appsv1beta1 "k8s.io/api/apps/v1beta1"
 	appsv1beta2 "k8s.io/api/apps/v1beta2"
@@ -75,6 +76,7 @@ var groups = []runtime.SchemeBuilder{
 	admissionv1.SchemeBuilder,
 	admissionregv1beta1.SchemeBuilder,
 	admissionregv1.SchemeBuilder,
+	apiserverinternalv1alpha1.SchemeBuilder,
 	appsv1beta1.SchemeBuilder,
 	appsv1beta2.SchemeBuilder,
 	appsv1.SchemeBuilder,
```

and generate fixtures with `UPDATE_COMPATIBILITY_FIXTURE_DATA=true go test k8s.io/api` for the generated commit

lgtm otherwise

liggitt commented:
/approve
for API file move

/hold
for [test data comment](https://github.com/kubernetes/kubernetes/pull/93087#issuecomment-666046074)

liggitt commented:
looking at the follow-up PR, I'm also confused as to whether kube-apiserver or kube-aggregator is the hosting server for this API, xref https://github.com/kubernetes/kubernetes/pull/92064#issuecomment-666050323

roycaihw commented:
updated testdata and squashed. Please take a look
```
commit 20f6784dd129438ca28b45c8f16939285ead56bc (HEAD -> storage-version/move-types)
Author: Haowei Cai <haoweic@google.com>
Date:   Thu Jul 30 10:06:29 2020 -0700

    fixup! generated

 staging/src/k8s.io/api/testdata/HEAD/internal.apiserver.k8s.io.v1alpha1.StorageVersion.json |  66 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 staging/src/k8s.io/api/testdata/HEAD/internal.apiserver.k8s.io.v1alpha1.StorageVersion.pb   | Bin 0 -> 350 bytes
 staging/src/k8s.io/api/testdata/HEAD/internal.apiserver.k8s.io.v1alpha1.StorageVersion.yaml |  46 ++++++++++++++++++++++++++++++++++++++++++++++
 3 files changed, 112 insertions(+)

commit cc587547599d67486d7e675d0b7acee1aee7b99f
Author: Haowei Cai <haoweic@google.com>
Date:   Thu Jul 30 10:06:10 2020 -0700

    fixup! fix import paths

 staging/src/k8s.io/api/roundtrip_test.go | 2 ++
 1 file changed, 2 insertions(+)
```

roycaihw commented:
> whether kube-apiserver or kube-aggregator is the hosting server for this API

The commit `Let kube-aggregator host the storage version API ` was from #84016. I cherrypicked @caesarxuchao 's commit and started from there. The host server is kube-apiserver

roycaihw commented:
/retest

liggitt commented:
/lgtm
/hold cancel

liggitt commented:
/test pull-kubernetes-e2e-kind-ipv6

roycaihw commented:
resolved the conflict in `hack/.golint_failures`

caesarxuchao commented:
/lgtm

caesarxuchao commented:
/lgtm
